transports/quic/src/transport.rs

```diff
@@ -223,14 +286,125 @@ impl<P: Provider> Transport for GenTransport<P> {
         }
 
         if let Poll::Ready(Some(ev)) = self.listeners.poll_next_unpin(cx) {
-            return Poll::Ready(ev);
+            match ev {
+                TransportEvent::Incoming {
+                    listener_id,
+                    upgrade,
```

Comment:
I don't think there is something bad happens if we change `ListenerUpgrade` from `Connecting` to `BoxFuture<'static, Result<Self::Output, Self::Error>>;`. If you do that, `GenTransport::poll` must poll `upgrade`s from `Listener`, you will get a `(PeerId, Connection)` you can work with. So you don't have to `poll` upgrade inside `HolePuncher::poll` anymore. That's the only way we can distinguish `HolePunchKey` by `(SocketAddr, PeerId)`.

And that's actually a good idea because one can `listen_on`, make a `dial_as_listener` to Aaddr:Aid but receive a valid incoming connection from Aaddr:Bid which will not be surprising at all.

Reply:
Also imagine A sends 2 dials (draft29 and v1) on a B listener (v1) that tries to punch that same client and draft29 comes first. B then stops punching but fails to complete handshake. It will result in error from `dial_as_listener` on B side.